File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(const string& s) {
        std::unordered_set<char> unique_chars;
        int longest = 0, l = 0, r = 0;

        while (r < s.length()) {
            if (!unique_chars.count(s[r])) {
                unique_chars.insert(s[r]);
                longest = std::max(longest, r - l + 1);
            } else {
                while (l < r && s[l] != s[r]) {
                    unique_chars.erase(s[l]);
                    l++;

                }
                l++;
            }
            r++;
        }

        return longest;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        unique_chars = set()
        l, longest = 0, 0

        for r in range(len(s)):
            if s[r] not in unique_chars:
                unique_chars.add(s[r])
                longest = max(longest, r - l + 1)
            else:
                while l < r and s[l] != s[r]:
                    unique_chars.remove(s[l])
                    l += 1
                l += 1
        
        return longest
        
